File: src/insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
# https://leetcode.com/problems/insert-delete-getrandom-o1

'''
One of those problems where you have to recognize the little trick,
and then it's pretty easy. Kind of a fun little gimmick. Swap the
intermediate element to pop with the last element in the list so we
can pop from end for constant time.
'''

import random

class RandomizedSet(object):

    def __init__(self):
        self.vals = list()
        self.indexmap = dict()

    def insert(self, val):
        """
        :type val: int
        :rtype: bool
        """
        if val in self.indexmap:
            return False
        idx = len(self.vals)
        self.vals.append(val)
        self.indexmap[val] = idx
        return True

    def remove(self, val):
        """
        :type val: int
        :rtype: bool
        """
        if val not in self.indexmap:
            return False
        # Swap element to delete with last element, and then
        # pop from end rather than pop from middle (O(1) vs O(N)).
        idx = self.indexmap[val]
        self.vals[-1], self.vals[idx] = self.vals[idx], self.vals[-1]
        self.indexmap[self.vals[idx]] = idx
        self.vals.pop()
        del self.indexmap[val]
        return True
        

    def getRandom(self):
        """
        :rtype: int
        """
        idx = random.randint(0, len(self.vals) - 1)
        return self.vals[idx]

if __name__ == "__main__":
    randset = RandomizedSet()
    ops = [
        ("insert",      (1,),   True),
        ("remove",      (2,),   False),
        ("insert",      (2,),   True),
        ("getRandom",   (),     "rand"),
        ("remove",      (1,),   True),
        ("insert",      (2,),   False),
        ("getRandom",   (),     2),
    ]
    for (method, args, out) in ops:
        method = getattr(randset, method)
        ret = method(*args)
        if out != "rand":
            assert out == ret, \
                f"out {out} != ret {ret}"
    print("✅ All tests passed")

File: src/insert-delete-getrandom-o1/insert-delete-getrandom-o1.rs
// https://leetcode.com/problems/insert-delete-getrandom-o1

use std::collections::HashMap;
use rand::rngs::ThreadRng;
use rand::seq::SliceRandom;

struct RandomizedSet {
    values: Vec<i32>,
    index: HashMap<i32, usize>,
    thread_rng: ThreadRng,
}

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RandomizedSet {
    fn new() -> Self {
        let values = vec![];
        let index = HashMap::new();
        let thread_rng = rand::thread_rng();

        Self { values, index, thread_rng }
    }
    
    fn insert(&mut self, val: i32) -> bool {
        if self.index.contains_key(&val) {
            return false;
        }

        self.index.insert(val, self.values.len());
        self.values.push(val);

        true
    }
    
    fn remove(&mut self, val: i32) -> bool {
        let idx = match self.index.get(&val) {
            None => return false,
            Some(idx) => *idx,
        };
        let last = self.values.last()
            .expect("Attempt to remove() while empty.")
            .to_owned();
        assert!(self.values.swap_remove(idx) == val);
        assert!(self.index.insert(last, idx).is_some());
        assert!(self.index.remove(&val).is_some());

        true
    }
    
    fn get_random(&mut self) -> i32 {
        self.values.as_slice()
            .choose(&mut self.thread_rng)
            .expect("Attempt to get_random() while empty")
            .to_owned()
    }

}

fn main() {
    let mut obj = RandomizedSet::new();
    assert!(obj.insert(1));
    assert!(obj.get_random() == 1);
    assert!(obj.remove(1));
    assert!(!obj.remove(2));

    assert!(obj.insert(0));
    assert!(obj.insert(1));
    assert!(obj.insert(2));
    for _ in 0..100 {
        let val = obj.get_random();
        assert!(val == 0 || val == 1 || val == 2)
    }

    assert!(obj.remove(1));
    assert!(!obj.remove(1));
    for _ in 0..100 {
        let val = obj.get_random();
        assert!(val == 0 || val == 2)
    }
    assert!(obj.remove(0));
    assert!(obj.remove(2));
 
    println!("✅ All tests passed");
}
